### PR TITLE
Add support for configuration of cache directory

### PR DESCRIPTION
**Pull request by adellibovi**
Resolves https://github.com/tuist/tuist/issues/2440
Request for comments document (if applies):

### Short description 📝

This PR is still a draft.

### Checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase.
- [x] The changes have been tested following the [documented guidelines](https://tuist.io/docs/contribution/testing-strategy/).
- [x] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [x] In case the PR introduces changes that affect users, the documentation has been updated.

### CODE REVIEW

File: Sources/TuistCache/GraphMappers/TestsCacheGraphMapper.swift
Comment (adellibovi) [inline Sources/TuistCache/GraphMappers/TestsCacheGraphMapper.swift:167]: @fortmarek is this change correct?

Comment (fortmarek) [reply Sources/TuistCache/GraphMappers/TestsCacheGraphMapper.swift:167]: I think so 👍

**Comment by fortmarek** [comment]
Hi, do you want me to look at the draft PR or is it better to wait until this is ready for proper review? @adellibovi

**Comment by adellibovi** [comment]
Hi @fortmarek, yes, if you can have already a look it would be great. In fact I have few concrete questions you may help with 😄 

* I created a CacheDirectoriesProvider under TuistCore, does it fit this module?
* As you can see we will still have "shared" cache, like `manifestCacheDirectory` and `projectDescriptionHelpersCacheDirectory`, as those are a requirement to get the Config where the cache configuration is. How `tuist clean` should behave when declaring a custom cache directory? Should it automatically clean both shared and local under the hood? This would be my suggested approach.
* I was thinking we could replace `TUIST_CACHE_DIRECTORY` environment variable used in the integrations test for the Config one. Thoughts? Or probably better, we could leave it for a different PR.

P.S.: I am still missing a couple of things: documentation, mapping of the config models and some tests.

File: Sources/TuistCore/Cache/CacheDirectoriesProvider.swift
Comment (luispadron) [inline Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:38]: nit: lets keep the docs in the protocol since less places to maintain.
```suggestion
    /// Returns the directory where the build artifacts are cached.
```

**Comment by fortmarek** [comment]
> I created a CacheDirectoriesProvider under TuistCore, does it fit this module?

Is there a reason why it can not be in `TuistCache`? That might be a better fit.

> As you can see we will still have "shared" cache, like manifestCacheDirectory and projectDescriptionHelpersCacheDirectory, as those are a requirement to get the Config where the cache configuration is. How tuist clean should behave when declaring a custom cache directory? Should it automatically clean both shared and local under the hood? This would be my suggested approach.

Should not be locations of `manifestCacheDirectory` and `projectDescriptionHelpersCacheDirectory` modified as well if user specifies it? I'd expect that as a user. 
If that'd mean too many changes, I think it's fine for this to be done in a follow-up PR - in the meantime I'd probably remove both but after all will be in a local cache dir, on `clean` I'd just remove the local cache dir.

> I was thinking we could replace TUIST_CACHE_DIRECTORY environment variable used in the integrations test for the Config one. Thoughts? Or probably better, we could leave it for a different PR.

Yes, I don't think that is necessary anymore since we will make it a part of the public API. It's fine to do those changes in this PR if that does not mean too many changes, I'll leave that to you to decide 🙂 

> Do you have any suggestion for the property under the Cache configuration? Currently it is just "path".

Not really 😄 `cacheLocation`, `cachePath` came to my mind but the `cache` prefix seems redundant. I'm fine with `path` personally since it's documented well.

File: Sources/ProjectDescription/Cache.swift
Comment (fortmarek) [inline Sources/ProjectDescription/Cache.swift:26]:
```suggestion
    /// The path where the cache will be stored, if `nil` it will be a default location in a shared directory.
```

File: Sources/ProjectDescription/Cache.swift
Comment (fortmarek) [inline Sources/ProjectDescription/Cache.swift:34]: nitpick: I'd change the comment the same way as above + I'd use `- Parameters: ` for documentation instead of writing twice `Parameter`

File: Sources/TuistCore/Cache/CacheDirectoriesProvider.swift
Comment (fortmarek) [inline Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:11]: afaict this is not used anywhere (probably a mistake by me). I think it's safe to delete this.

File: Sources/TuistCoreTesting/Cache/MockCacheDirectoriesProvider.swift
Comment (fortmarek) [inline Sources/TuistCoreTesting/Cache/MockCacheDirectoriesProvider.swift:6]: nitpick: I'd make this `final`

File: Sources/TuistCoreTesting/Cache/MockCacheDirectoriesProvider.swift
Comment (fortmarek) [inline Sources/TuistCoreTesting/Cache/MockCacheDirectoriesProvider.swift:7]: nitpick: I'd make the properties `private`

File: Sources/TuistCore/Cache/CacheDirectoriesProvider.swift
Comment (fortmarek) [inline Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:20]: nit: I'd make this `final`

File: Sources/TuistKit/Services/TestService.swift
Comment (fortmarek) [inline Sources/TuistKit/Services/TestService.swift:65]: to be more consistent with the codebase, I'd create a new component `final class CacheDirectoryProviderFactory` instead of a closure.

**Review by fortmarek — COMMENTED** [review]
Love to see this, I'm excited to use the functionality myself 🎉 I have added some minor comments but the PR is in the right direction 🙂 good job @adellibovi!

File: Sources/TuistCore/Cache/CacheDirectoriesProvider.swift
Comment (adellibovi) [reply Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:21]: Currently, it is because of the clean service, as it is used for getting the "shared cache location" too

File: Sources/TuistSupport/Utils/Environment.swift
Comment (adellibovi) [reply Sources/TuistSupport/Utils/Environment.swift:75]: Agree, but I can't see how `manifestCacheDirectory` and `projectDescriptionHelpersCacheDirectory` moved the the local cache, I will give it a thought :)

**Comment by adellibovi** [comment]
Thanks for the inputs @fortmarek!!!
I already applied most of your suggestions, and left 2 questions opens for now: optionality for Config which depends on clean command and the `manifestCacheDirectory` and `projectDescriptionHelpersCacheDirectory` comment which is also somehow related. I will review them more deeply and keep you updated 😄

File: Sources/TuistCore/Cache/CacheDirectoriesProvider.swift
Comment (fortmarek) [inline Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:33]: I think we can delete this variable. It was created for acceptance tests purposes but since we'll be able to define this via public API, it should no longer be necessary

File: Sources/TuistKit/ProjectEditor/ProjectEditor.swift
Comment (fortmarek) [inline Sources/TuistKit/ProjectEditor/ProjectEditor.swift:96]: I don't think this is correct as we will use the real implementation of helpers builders in unit tests

File: Sources/TuistLoader/Loaders/CachedManifestLoader.swift
Comment (fortmarek) [inline Sources/TuistLoader/Loaders/CachedManifestLoader.swift:107]: factory used here should be also injected

File: Sources/TuistLoader/Loaders/CachedManifestLoader.swift
Comment (fortmarek) [inline Sources/TuistLoader/Loaders/CachedManifestLoader.swift:106]: we should run this only if the `manifest` is `.config`, shouldn't we? Right now this is run whenever `cacheDirectory` is nil

**Review by fortmarek — COMMENTED** [review]
This is in good direction @adellibovi! Might be worth updating the fixture, so it's possible to try out the changes locally 🙂

File: Sources/TuistCore/Cache/CacheDirectoriesProvider.swift
Comment (adellibovi) [reply Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:40]: good point

File: Sources/TuistLoader/Loaders/CachedManifestLoader.swift
Comment (adellibovi) [reply Sources/TuistLoader/Loaders/CachedManifestLoader.swift:106]: The goal here is to load the config so we can set the cacheDirectory for the other type of manifests. Maybe if it is not clear I could try to extract as a lazy var? or do you have other suggestions? 🤔

Comment (adellibovi) [reply Sources/TuistLoader/Loaders/CachedManifestLoader.swift]: Actually, this part does not seem to be correct, as the path is the root of others manifest. Do you have ideas on how I can find where Config is located?

Comment (fortmarek) [reply Sources/TuistLoader/Loaders/CachedManifestLoader.swift]: yes, exactly!

File: Sources/TuistKit/ProjectEditor/ProjectEditor.swift
Comment (adellibovi) [reply Sources/TuistKit/ProjectEditor/ProjectEditor.swift]: fixed

File: Sources/TuistLoader/Loaders/CachedManifestLoader.swift
Comment (adellibovi) [reply Sources/TuistLoader/Loaders/CachedManifestLoader.swift]: fixed

**Comment by adellibovi** [comment]
Hello @fortmarek,

I updated the PR, hopefully it is now ready to be moved forward. 🙏 
I would suggest to still keep the environment variable for forcing the cache directory in order to not increase the amount of changes further, though, I did move it at the "Config" level, so effectively now it behaves as it would be configured in the Config.swift, following the new option introduced with this PR.

There are other issues happening in the `main` branch and not sure if there are known issues. cc: @tuist/core
* Acceptance tests are not running in `main` branch, it fails to find steps definitions, but it still results CI being green.
* `tuist cache warm` + `tuist focus` is not working in `main` branch, acceptance tests (once correctly executed) will pass, but only because the cache directory is forced to be under the source root (to be specific, during test execution, it's the same directory as the temporary source directory). Instead, when executing tuist normally, the default cache directory is under `~/.tuist/Cache` and header search path cannot reference frameworks outside the project source in the form of `$SRCROOT/../../../../~/.tuist/Cache`.

I would suggest to fix the `tuist cache warm` + `tuist focus` issue in a different PR.
This PR will improve the situation by introducing a workaround for the users as they can now define a custom cache directory, which can be under the source root path, which will work properly.

File: Sources/TuistSupport/Constants.swift
Comment (fortmarek) [inline Sources/TuistSupport/Constants.swift:57]: I'd mark this deprecated and add a link to the issue, so it's obvious that this will be used only temporarily 'till we fix it

**Review by fortmarek — COMMENTED** [review]
Looking good, I only have some smaller comments, looking forward to merging this 👀

File: Sources/TuistLoader/Loaders/CachedManifestLoader.swift
Comment (adellibovi) [reply Sources/TuistLoader/Loaders/CachedManifestLoader.swift:26]: I was thinking that this property is more like a late init one, when used we should always have a value, if we access prematurely it's a logic error. If we convert to optional we have a path where CachedManifestLoader won't use the cache and shouldn't not be the case, which in my opinion feel even stranger 🤔

File: Sources/TuistCache/Cache/Cache.swift
Comment (adellibovi) [reply Sources/TuistCache/Cache/Cache.swift:17]: I noticed that errors in this file already fails silently, that's why I didn't propagate.
Good point about init-work, I moved to a computed property :)

File: Sources/TuistCore/Cache/CacheDirectoriesProvider.swift
Comment (adellibovi) [reply Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:37]: 😛

File: Sources/TuistLoader/Loaders/CachedManifestLoader.swift
Comment (adellibovi) [reply Sources/TuistLoader/Loaders/CachedManifestLoader.swift:107]: ah! Nice one

File: Sources/TuistLoader/Loaders/CachedManifestLoader.swift
Comment (adellibovi) [reply Sources/TuistLoader/Loaders/CachedManifestLoader.swift:201]: It was more to explicitly pass the unwrapped property, anyway changed :)

File: Sources/TuistPlugin/PluginService.swift
Comment (adellibovi) [reply Sources/TuistPlugin/PluginService.swift:48]: Done

File: Sources/TuistSupport/Constants.swift
Comment (adellibovi) [reply Sources/TuistSupport/Constants.swift:57]: done

File: Sources/TuistCore/Cache/CacheDirectoriesProvider.swift
Comment (pepibumur) [inline Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:6]: I'd differentiate between the caching feature of Tuist, and the internal caches that Tuist uses to speed up project generation. That said, I'd rename this to something more generic, `ProjectDirectoriesProviding` and `ProjectDirectoriesProvider`. Given a project configuration and its path, the instance returns paths to several directories.

Comment (pepibumur) [inline Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:6]: Does it need to inherit from `AnyObject`?

File: Sources/TuistCore/Cache/CacheDirectoriesProvider.swift
Comment (pepibumur) [inline Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:17]: Nitpick: `generatedAutomationProjectsDirectory`. It's longer, but more explicit about what the directory is for.

File: Sources/TuistCore/Cache/CacheDirectoriesProvider.swift
Comment (pepibumur) [inline Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:36]: Side effects in class constructors are usually not a great idea. As a developer consuming `CacheDirectoriesProvider`, I'd not expect IO operations to happen at initialization time. I'd rather remove the logic from here, and let the entity that needs to use that directory make sure it's in the right state.

File: Sources/TuistCore/Cache/CacheDirectoriesProvider.swift
Comment (pepibumur) [inline Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:31]: As I mentioned in another comment, Tuist internal caching and caching of the project modules are not the same thing and therefore we should make sure we differentiate them. I'd use `projectBinariesCachePath` to refer to `config.cache.path`, and `tuistCachePath`, to refer to the cache that Tuist uses internally. 
Using very explicit names will avoid confusion down the road.

File: Sources/TuistKit/Cache/CacheStorageProvider.swift
Comment (pepibumur) [inline Sources/TuistKit/Cache/CacheStorageProvider.swift:14]: After looking at the factory pattern, I thought you'd use it for doing dependency injection and change the instance of the directories provider. However, you are not doing any dependency injection. Was that the goal? If we can easily test the classes without that, do we really need the factory?

**Review by pepibumur — COMMENTED** [review]
👋 Thanks a lot for doing this @adellibovi. Sooner or later we'd need a class like the one that you introduced to abstract away the logic that constructs the paths that are used by different Tuist features. As I posted in one of my comments, I'd change the name of that class to be more generic, and make sure we differentiate with more explicit naming what's the Tuist internal cache and what's the cache for storing project's binaries. Once those renames have been done, I'll continue with the review of the PR.

File: Sources/TuistCore/Cache/CacheDirectoriesProvider.swift
Comment (adellibovi) [reply Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:36]: I agree @pepibumur. The code here was moved from Environment.swift (where the existence and setup happens on init). The init in Environment was not throwing because it was using a force unwrap, based on the comment of @fortmarek I change this one here for throwing. If we move this to the computed property we would either move back to use force unwrap or convert all the properties to throwing methods.
Should I still try to change in this PR?

File: Sources/TuistCore/Cache/CacheDirectoriesProvider.swift
Comment (adellibovi) [reply Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:17]: This was all existing naming, yet I am happy to change it :)

File: Sources/TuistCore/Cache/CacheDirectoriesProvider.swift
Comment (adellibovi) [reply Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:6]: Not needed, changed

File: Sources/TuistCore/Cache/CacheDirectoriesProvider.swift
Comment (fortmarek) [reply Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:36]: I think we can a) let a different component / user of this component make sure that `cacheDirectory` exists b) create a function along the lines of `setupCache()` that will create it.

Comment (adellibovi) [reply Sources/TuistCore/Cache/CacheDirectoriesProvider.swift:36]: As commented offline with Marek, this was moved a level up in `CacheDirectoriesProviderFactory` so that it can create `CacheDirectoriesProvider` and make sure the directories are ready to be used or created otherwise :)

**Comment by adellibovi** [comment]
Hi @pepibumur and @fortmarek CI seem to be green now :)

**Comment by fortmarek** [comment]
> CI seem to be green now :)

it seems to be red actually, although, only the webpage checks, so not sure if this is related to your changes 🤔 could you maybe try to rebase and see if it passes then?

**Review by fortmarek — APPROVED** [review]

**Comment by adellibovi** [comment]
> > CI seem to be green now :)
> 
> it seems to be red actually, although, only the webpage checks, so not sure if this is related to your changes 🤔 could you maybe try to rebase and see if it passes then?

@fortmarek thanks, the branch is/was already in sync with main branch, I am not sure what's going on, but this is what it returns from the log of Github Action:
```
5:04:24 PM: [1/3] Compiling website main.swift
5:04:24 PM: <unknown>:0: error: PCH was compiled with module cache path '/opt/build/repo/website/.build/x86_64-unknown-linux-gnu/debug/ModuleCache/26F2SGHBM33EN', but the path is currently '/opt/build/repo/projects/website/.build/x86_64-unknown-linux-gnu/debug/ModuleCache/26F2SGHBM33EN'
5:04:24 PM: <unknown>:0: error: missing required module 'SwiftShims'
5:04:24 PM: Error building Swift package
```
It seems to me that is somehow related to caching of Netfly and the project directory changes? cc @pepibumur 
Maybe we can try cleaning the cache?

**Pull request merged**
